# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/systemservices/objc_video_capture/video_capture_avfoundation.h
# Commit: fe7300653e99
# Full Hash: fe7300653e993df24d32d69dc0efdc5cbb6f4c76
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2023-03-30 15:09:05
# Regressor Bug: 1817724
# File Overlap Count: 1
# Description:
#   Bug 1817724 - Don't buffer frames in VideoCaptureAvFoundation. r=webrtc-reviewers,ng
#   
#   This patch uses StateWatching to dispatch a task to process a frame as soon as
#   it comes in from the VideoCaptureAdapter frame event. This is done async so that
#   if multiple frames come in while one is being processed, we drop all but the
# ==============================================================================

diff -r 9aa28e73aaf8 -r fe7300653e99 dom/media/systemservices/objc_video_capture/video_capture_avfoundation.h
--- a/dom/media/systemservices/objc_video_capture/video_capture_avfoundation.h	Thu Mar 30 10:36:33 2023 +0000
+++ b/dom/media/systemservices/objc_video_capture/video_capture_avfoundation.h	Thu Mar 30 10:36:34 2023 +0000
@@ -14,6 +14,7 @@
 #include "MediaEventSource.h"
 #include "modules/video_capture/video_capture_impl.h"
 #include "mozilla/Maybe.h"
+#include "mozilla/StateWatching.h"
 #include "PerformanceRecorder.h"
 
 @interface VideoCaptureAdapter : NSObject <RTCVideoCapturerDelegate> {
@@ -54,11 +55,19 @@
   void SetTrackingId(uint32_t aTrackingIdProcId) override;
 
  private:
+  // Convert mNextFrameToProcess and pass it to callbacks.
+  void ProcessNextFrame();
+
   // Control thread checker.
   SequenceChecker mChecker;
+  const RefPtr<mozilla::AbstractThread> mCallbackThread;
   AVCaptureDevice* _Nonnull const mDevice RTC_GUARDED_BY(mChecker);
   VideoCaptureAdapter* _Nonnull const mAdapter RTC_GUARDED_BY(mChecker);
   RTCCameraVideoCapturer* _Nonnull const mCapturer RTC_GUARDED_BY(mChecker);
+  mozilla::WatchManager<VideoCaptureAvFoundation> mWatchManager RTC_GUARDED_BY(mChecker);
+  // The next frame to be processed. If processing is slow this is always the most recent frame.
+  mozilla::Watchable<__strong RTCVideoFrame* _Nullable> mNextFrameToProcess
+      RTC_GUARDED_BY(mChecker);
   // If capture has started, this is the capability it was started for.
   mozilla::Maybe<VideoCaptureCapability> mCapability RTC_GUARDED_BY(mChecker);
   // The image type that mCapability maps to. Set in lockstep with mCapability.